The New Github Auth Flow Blog Post (#469)
tinacms/tinacms.org: content/blog/the-new-github-auth-flow.md

---

> title: The New Github Auth Flow

Reviewer: ```suggestion
title: The New GitHub Auth Flow
```

> title: The New GitHub Auth Flow

Reviewer: ```suggestion
title: Upgrade Notice: TinaCMS + GitHub Packages
```
Maybe this instead :sweat_smile:

Reviewer 2: Filename too? Maybe `upgrade-notice-github-auth.md`

date: '2020-06-16T13:47:03-03:00'
author: Joel Huggett
---

> TinaCMS communicates with Github using a proxy, so the authentication token provided by Github is stored as an httpOnly cookie. This stops the client from accessing the token, and that's all very good. **But** what it doesn't account for is [CSRF]() attacks, which means any calls to the proxy, so long as that cookie is still there, will succeed, and that's not very good.

Reviewer: ```suggestion
TinaCMS communicates with GitHub using a proxy, so the authentication token provided by GitHub is stored as an httpOnly cookie. This stops the client from accessing the token, and that's all very good. However, this strategy is still vulnerable to [Cross-Site Request Forgery (CSRF)](https://owasp.org/www-community/attacks/csrf) attacks. This means that any calls to the proxy, so long as that cookie is still there, will succeed, and that's not very good.
```

> A common approach to mitigating this problem is to implement the [Token Synchronization Pattern](). The issue is that this pattern require some form of server-side session storage. That doesn't jive well with the stateless approach of static sites. So, we've introduced a variation that we call the Stateless Token Synchronization Pattern.

Reviewer: link missing

> It works by storing a CSRF token as an httpOnly cookie and sending an encrypted (signed by the server's secret _Signing Key_) token that is the amalgamation of the CSRF token and the authentication token provided by Github. This amalgamated token is then stored client-side in local storage and sent to the proxy in a bearer authentication header. Then, server-side, the amalgamated token is decrypted and the CSRF tokens are compared to make sure they match. If all is well, the authentication token is extracted and the call is completed.

Reviewer: ```suggestion
**Stateless Token Synchronization** works by storing a CSRF token as an httpOnly cookie and sending an encrypted (signed by the server's secret _Signing Key_) token that is the amalgamation of the CSRF token and the authentication token provided by Github. This amalgamated token is then stored client-side in local storage and sent to the proxy in a bearer authentication header. Then, server-side, the amalgamated token is decrypted and the CSRF tokens are compared to make sure they match. If all is well, the authentication token is extracted and the call is completed.
```

This new pattern help mitigate CSRF attacks and provides the authentication token in an encrypted format, all done statelessly.

> # Upgrading to the new flow

Reviewer: ```suggestion
## Upgrading to the new flow
```

### **react-tinacms-github**

Nothing needs to be changed. This package can handle both the old flow and the new one on its own without further configuration.

### **next-tinacms-github**

**next-tinacms-github** api routes now require a secret _Signing Key_.

> The _Signing Key_ is just a random passphrase of your choice, a randomized 256-bit key is recommended. This key is only used server-side to encrypt and decrypt authentication tokens sent to the client.

Reviewer: ```suggestion
The _Signing Key_ should be a random 256-bit key, used server-side to encrypt and decrypt authentication tokens sent to the client.
```

Reviewer: I would like to be a little more opinionated here, just to discourage people from making a 'weak' key. This isn't a super important decision for a dev to make so let's just tell them what they should do.

Reviewer 2: > ~should be~ **is** a random

> One way to generate a recommended size key, is to run `openssl rand -base64 32` in your terminal and use the output as your _Signing Key_.

Reviewer: ```suggestion
You can generate a key by running `openssl rand -base64 32` in your terminal, using the output as your _Signing Key_.
```

Reviewer: Again, just trying to be a tiny bit more forceful in the language here. The sentiment I'm going for is "just do it this way unless you have a really good reason"

`createAuthHandler`, `apiProxy`, and `previewHandler` now  **require** the _Signing Key_ to be passed as a parameter.

E.g., the proxy-github route:

```TSX
import { apiProxy } from 'next-tinacms-github'

export default apiProxy(
    process.env.SIGNING_KEY
)

> ```

Reviewer 2: It might be good to show all three examples, use the `diff` code block, and reference the names of the files that are created in the open authoring guide
for example...

**pages/api/proxy-github.ts**
```diff
import { apiProxy } from 'next-tinacms-github'

- export default apiProxy
+ export default apiProxy(
+   process.env.SIGNING_KEY
+ )

Reviewer: agreed, let's add explicit examples for every file that needs to be changed
